src/IceRpc.Protobuf.Generators/Internal/Emitter.cs

```diff
@@ -19,26 +19,40 @@ internal string Emit(IReadOnlyList<ServiceClass> serviceClasses, CancellationTok
             string dispatchImplementation;
             if (serviceClass.ServiceMethods.Count > 0)
             {
+                methodModifier += " async";
+
                 dispatchImplementation = "";
                 foreach (ServiceMethod serviceMethod in serviceClass.ServiceMethods)
                 {
-                    string operationName = serviceMethod.OperationName;
-                    string dispatchMethodName = serviceMethod.DispatchMethodName;
-                    dispatchImplementation +=
-                        $"\"{operationName}\" => global::{dispatchMethodName}(this, request, cancellationToken),\n";
+                    dispatchImplementation += @$"
+case ""{serviceMethod.OperationName}"":
+{{
+    var inputParam = new {serviceMethod.InputTypeName}();
+    await inputParam.MergeFromAsync(request.Payload).ConfigureAwait(false);
+    var outputParam = await (this as {serviceMethod.InterfaceName}).{serviceMethod.MethodName}(
+        inputParam,
+        request.Features,
+        cancellationToken).ConfigureAwait(false);
+    return new IceRpc.OutgoingResponse(request)
+    {{
+        Payload = outputParam.ToPipeReader()
+    }};
+}}".Trim();
                 }
-
                 if (serviceClass.HasBaseServiceClass)
                 {
-                    dispatchImplementation += "_ => base.DispatchAsync(request, cancellationToken)";
+                    dispatchImplementation += @$"
+default:
+    return base.DispatchAsync(request, cancellationToken);";
                 }
                 else
                 {
-                    dispatchImplementation +=
-                        "_ => new(new IceRpc.OutgoingResponse(request, IceRpc.StatusCode.NotImplemented))";
+                    dispatchImplementation += @$"
+default:
+    return new IceRpc.OutgoingResponse(request, IceRpc.StatusCode.NotImplemented);";
```

Comment:
I think there is a small bug here: should be `new(new...)` when not async. Let's fix it when we add more tests.